graphql_compiler/compiler/ir_lowering_common.py

```diff
@@ -1,24 +1,26 @@
 # Copyright 2017 Kensho Technologies, Inc.
 """Language-independent IR lowering and optimization functions."""
 
-from funcy.py2 import pairwise
-
-from .blocks import ConstructResult, Filter
+from .blocks import ConstructResult, Filter, Fold, Unfold
 from .expressions import (BinaryComposition, ContextField, ContextFieldExistence, FalseLiteral,
                           NullLiteral, TrueLiteral)
 from .helpers import validate_safe_string
 
 
 def merge_consecutive_filter_clauses(ir_blocks):
     """Merge consecutive Filter(x), Filter(y) blocks into Filter(x && y) block."""
+    if not ir_blocks:
+        return ir_blocks
+
     new_ir_blocks = [ir_blocks[0]]
 
-    for previous_block, current_block in pairwise(ir_blocks):
-        if isinstance(previous_block, Filter) and isinstance(current_block, Filter):
+    for block in ir_blocks[1:]:
+        last_block = new_ir_blocks[-1]
+        if isinstance(last_block, Filter) and isinstance(block, Filter):
             new_ir_blocks[-1] = Filter(
```

Comment:
I'm sure you're aware, but Python has shallow copies, and if `Filter(..)` here changes the pointer that `new_ir_blocks[-1]` points to, then we lose whatever there is. Are we sure we're overwriting here ?

Reply:
`CompilerEntity` and all its subclasses (including these ones) are meant to be immutable, though Python doesn't have great ways of enforcing this. All the compiler's code creates new blocks and new expressions instead of editing them.

Reply:
Good to know that `CompilerEntity` is enforcing immutability.

Reply:
It's not enforced (grumble grumble Python), but this is the intent of the design.